private/bufpkg/bufpluginexec/wasm_handler.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufpluginexec
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufwasm"
+	"github.com/bufbuild/buf/private/pkg/app"
+	"github.com/bufbuild/buf/private/pkg/app/appproto"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/types/pluginpb"
+)
+
+type wasmHandler struct {
+	wasmPluginExecutor bufwasm.PluginExecutor
+	pluginPath         string
+	tracer             trace.Tracer
+}
+
+func newWasmHandler(
+	wasmPluginExecutor bufwasm.PluginExecutor,
+	pluginPath string,
+) *wasmHandler {
+	return &wasmHandler{
+		wasmPluginExecutor: wasmPluginExecutor,
+		pluginPath:         pluginPath,
+		tracer:             otel.GetTracerProvider().Tracer("bufbuild/buf"),
+	}
+}
+
+func (h *wasmHandler) Handle(
+	ctx context.Context,
+	container app.EnvStderrContainer,
+	responseWriter appproto.ResponseBuilder,
+	request *pluginpb.CodeGeneratorRequest,
+) error {
+	ctx, span := h.tracer.Start(ctx, "plugin_proxy", trace.WithAttributes(
+		attribute.Key("plugin").String(filepath.Base(h.pluginPath)),
+	))
+	defer span.End()
+
+	path, err := validateWASMFilePath(h.pluginPath)
+	if err != nil {
+		span.RecordError(err)
```

Comment:
nit: It makes it a lot easier to ensure spans are marked appropriately if we use a named error and a defer.

Reply:
Do we not also want to `span.SetStatus(codes.Error, err.Error())` here?